### PR TITLE
#163789190 Input Validation Function

### PR DESCRIPTION
**Pull request by eokwukwe**
#### What does this PR do?
Create a universal input validation function
#### Description of Task to be completed?
- [x] write a unit test for input validation function
- [x] create a universal input validation function
- [x] create a sign in schema and test
- [x] create a sign-up schema and test
#### How should this be manually tested?
- [x] clone the repo
- [x] cd into the folder and run `npm install` to install dependencies
- [x] run `npm run test`
#### What are the relevant pivotal tracker stories?
[#163789190](https://www.pivotaltracker.com/story/show/163789190)

### CODE REVIEW

**Review by ODINAKACHUKWU — APPROVED** [review]

**Comment by benfluleck** [comment]
Look at your commit messages and change them

File: server/middlewares/validations/user.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import Joi from "joi";
+
+const name = Joi.string().min(2);
```
Comment (benfluleck) [inline server/middlewares/validations/user.js]: Don't we want only alphabets in the name

File: server/middlewares/validations/user.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  .min(7)
+  .error(errors => {
+    return {
+      message:
+        "password must be atleast 6 chars with atleast 1 uppercase, 1 number, & 1 special char"
+    };
+  });
+
+const createUserSchema = Joi.object().keys({
+  firstname: name.required(),
```
Comment (benfluleck) [inline server/middlewares/validations/user.js]: You can chain the methods on like `Joi.string().required().min(2)`

File: server/tests/index.spec.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
 import "./miscellaneous/connection.spec";
 import "./model.specs/userModel.specs";
 import "./model.specs/articles.specs";
+import "./validation.specs/user.spec";
```
Comment (benfluleck) [inline server/tests/index.spec.js]: the folder should just be called validations

File: server/tests/index.spec.js [hunk 4]
```diff
@@ -1,3 +1,4 @@
 import "./miscellaneous/connection.spec";
 import "./model.specs/userModel.specs";
 import "./model.specs/articles.specs";
```
Comment (benfluleck) [inline server/tests/index.spec.js]: Change the model.specs to model

File: server/tests/validation.specs/user.spec.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+import sinonChai from "sinon-chai";
+import user from "../../middlewares/validations/user";
+
+const expect = chai.expect;
+chai.use(sinonChai);
+
+describe("User input validation Middleware", () => {
+  it("should not return an error when all the inputs are valid", done => {
+    const value = {
+      firstname: "mss",
```
Comment (benfluleck) [inline server/tests/validation.specs/user.spec.js]: Define this at the top of the tests and re-use

File: server/tests/validation.specs/user.spec.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    };
+    let res = {};
+    user.validateUserInput(req, res, () => {
+      done();
+    });
+  });
+
+  it("should return an error when firstname is invalid", done => {
+    const value = {
+      firstname: "",
```
Comment (benfluleck) [inline server/tests/validation.specs/user.spec.js]: Keep your logic DRY

File: server/tests/validation.specs/user.spec.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        expect(statusSpy).to.have.been.calledOnce;
+        done();
+      })
+      .catch(error => console.log("error", error));
+  });
+
+  it("should return an error when lastname is invalid", done => {
+    const value = {
+      firstname: "john",
+      lastname: "",
```
Comment (benfluleck) [inline server/tests/validation.specs/user.spec.js]: Keep your Logic DRY

File: server/tests/validation.specs/user.spec.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        };
+      }
+    };
+    let statusSpy = sinon.spy(res, "status");
+
+    user
+      .validateUserInput(req, res, () => {})
+      .then(response => {
+        statusSpy.restore();
+        expect(statusSpy).to.have.been.calledOnce;
```
Comment (benfluleck) [inline server/tests/validation.specs/user.spec.js]: This is not enough testing, the response object has a lot more things we can test e.g. status code

File: server/tests/validation.specs/user.spec.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
```
Comment (benfluleck) [inline server/tests/validation.specs/user.spec.js]: You need to test a little more here

File: server/tests/validation.specs/user.spec.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
+import mocha from "mocha";
```
Comment (benfluleck) [inline server/tests/validation.specs/user.spec.js]: You are not using this line for anything?

File: server/middlewares/validations/user.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+const validateUserInput = async (req, res, next) => {
+  try {
+    const value = req.body;
+    await Joi.validate(value, createUserSchema, {
+      abortEarly: false
+    });
+    next();
+  } catch (error) {
+    const errorMessages = error.details.map(errorMsg => {
+      return errorMsg.message.replace(/"/g, "");
```
Comment (benfluleck) [inline server/middlewares/validations/user.js]:
```
catch (errors) {
errors.details.map(error => {
      return error.message.replace(/"/g, "");
 }
```

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/index.js [hunk 12]
```diff
@@ -51,7 +51,7 @@ app.use((error, req, res, next) => {
   next();
 });
 
-// error handler
+// development error handler, will print stacktrace
```
Comment (benfluleck) [inline server/index.js]: Line 54 is a duplication of 56 and 57

File: server/middlewares/validations/user.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+});
+
+/**
+ * Validate User Input on Registration
+ * @param {object} req The request body object
+ * @param {object} res
+ * @param {function} next
+ * @returns {object} Response object
+ */
+const validateUserInput = async (req, res, next) => {
```
Comment (benfluleck) [inline server/middlewares/validations/user.js]: I feel this logic can be re-used by other validation schemas, please refactor this;

File: server/tests/validations/user.spec.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+import sinon from "sinon";
+import sinonChai from "sinon-chai";
+import user from "../../middlewares/validations/user";
+
+const expect = chai.expect;
+chai.use(sinonChai);
+
+describe("User input validation Middleware", () => {
+  let value = {
+    firstname: "john",
```
Comment (benfluleck) [inline server/tests/validations/user.spec.js]: This should be taken from fixtures

File: server/tests/validations/user.spec.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    value = {};
+    req = {
+      body: JSON.stringify(value)
+    };
+
+    user
+      .validateUserInput(req, res, () => {})
+      .then(response => {
+        expect(response.status).to.equal("error");
+        statusSpy.restore();
```
Comment (benfluleck) [inline server/tests/validations/user.spec.js]: I want to see each error message logged.

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/middlewares/validations/user.js (hunk 13 above)
Comment (eokwukwe) [reply server/middlewares/validations/user.js]: I will need to get more clarification on this, please.

Comment (benfluleck) [reply server/middlewares/validations/user.js]: Certainly, the way this has been defined you have tightly coupled it with the createUserSchema. So now If I am writing another schema e.g. createArticle I would need a separate function for that as well. I need the logic to be reusable so any Joischema can use it.

Comment (eokwukwe) [reply server/middlewares/validations/user.js]: Okay. It means that the method name and the file name will also need to change to signify that it is a generic method.
Also, I intend to create a `validationSchema` folder in the `middlewares` that will hold all the schemas for the project.

**Review by obasajujoshua31 — APPROVED** [review]

File: package.json [hunk 16]
```diff
@@ -43,7 +43,9 @@
     "@sendgrid/mail": "^6.3.1",
     "cors": "^2.8.4",
     "express": "^4.16.3",
-    "jsonwebtoken": "^8.4.0",
+    "express-session": "^1.15.6",
```
Comment (benfluleck) [inline package.json]: Why is this package in here please remove it

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: package.json [hunk 17]
```diff
@@ -43,7 +43,9 @@
...
+    "joi": "^14.3.1",
+    "jsonwebtoken": "^8.3.0",
```
Comment (benfluleck) [inline package.json]: Why is this package just being installed are you sure you rebased properly?

File: server/tests/validations/user.spec.js [hunk 18]
```diff
@@ -0,0 +1,63 @@
+import chai from "chai";
+import { validateInput } from "../../validations/validationInput";
+import userData from "../fixtures/model/userData";
+import { signUpSchema } from "../../validationSchemas/user";
+
+const expect = chai.expect;
```
Comment (benfluleck) [inline server/tests/validations/user.spec.js]: destructure this please.

File: server/tests/validations/user.spec.js [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+describe("User input validation Middleware", () => {
```
Comment (benfluleck) [inline server/tests/validations/user.spec.js]: Is your validation middleware?

**Review by Ndifreke — APPROVED** [review]

**Comment by benfluleck** [comment]
Please implement signup schema as well so it will give people a better sense of how to use your validation. Thanks 😄

**Comment by benfluleck** [comment]
screenshot 2019-02-13 at 9 37 20 am
Why is my commit still in your workflow. Please find out why it is there, Also sort out your commit message. Make sure you are following convention.  Thanks @eokwukwe

File: server/configs/passport.js [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    ? {
+        clientID: FACEBOOK_CLIENT_ID_PROD,
+        clientSecret: FACEBOOK_CLIENT_SECRET_PROD,
+        callbackURL: FACEBOOK_REDIRECT_PROD,
+        profileFields: ["id", "displayName", "photos", "email"]
+      }
+    : {
+        clientID: FACEBOOK_CLIENT_ID,
+        clientSecret: FACEBOOK_CLIENT_SECRET,
+        callbackURL: `http://localhost:${PORT}/api/v1/auth/facebook/redirect`,
```
Comment (benfluleck) [inline server/configs/passport.js:28]: Isn't this commit from @obasajujoshua31 🤔

**Pull request closed without merging**